libraries/chain/block_log.cpp

```diff
@@ -469,20 +478,219 @@ namespace eosio { namespace chain {
       return reverse_block_position_iterator<BlockLogData>(t, first_block_position);
    }
+
+   template <typename Lambda>
+   void for_each_file_in_dir_matches(const fc::path& dir, const char* pattern, Lambda&& lambda) {
+      const std::regex my_filter(pattern);
+      std::smatch      what;
+      boost::filesystem::directory_iterator end_itr; // Default ctor yields past-the-end
+      for (boost::filesystem::directory_iterator p(dir); p != end_itr; ++p) {
+         // Skip if not a file
+         if (!boost::filesystem::is_regular_file(p->status()))
+            continue;
+         // skip if it's not match blocks-*-*.log
+         if (!std::regex_match(p->path().filename().string(), what, my_filter))
+            continue;
+         lambda(p->path());
+      }
+   } 
    } // namespace
 
-   block_log::block_log(const fc::path& data_dir)
-       : my(new detail::block_log_impl(data_dir)) {
-   }
+   struct block_log_catalog {
+      using block_num_t = uint32_t;
+
+      struct mapped_type {
+         block_num_t             last_block_num;
+         boost::filesystem::path filename_base;
+      };
+      using collection_t              = boost::container::flat_map<block_num_t, mapped_type>;
+      using size_type                 = collection_t::size_type;
+      static constexpr size_type npos = std::numeric_limits<size_type>::max();
+
+      boost::filesystem::path    archive_dir;
+      size_type                  max_retained_files = 10;
+      collection_t               collection;
+      size_type                  active_index = npos;
+      block_log_data             log_data;
+      block_log_index            log_index;
+      chain_id_type              chain_id;
+
+      bool empty() const { return collection.empty();  }
+
+      void open(fc::path block_dir) {
+         for_each_file_in_dir_matches(block_dir, R"(blocks-\d+-\d+\.log)", [this](boost::filesystem::path path) {
+            auto log_path               = path;
+            auto index_path             = path.replace_extension("index");
+            auto path_without_extension = log_path.stem().string();
+
+            block_log_data log(log_path);
+
+            if (chain_id.empty()) {
+               chain_id = log.chain_id();
+            } else {
+               EOS_ASSERT(chain_id == log.chain_id(), block_log_exception, "block log file ${path} has a different chain id",
+                           ("path", log_path.generic_string()));
+            }
+
+            // check if index file matches the log file
+            if (!index_matches_data(index_path, log))
+               block_log::construct_index(log_path, index_path);
+
+            auto [itr, _]      = collection.emplace(log.first_block_num(),
+                                                    mapped_type{ log.last_block_num(), path_without_extension });
+            this->active_index = collection.index_of(itr);
```

Comment:
If I'm not off base about this we should probably extend the test cases to cover situations where lexicographically ordered filenames are not properly ordered by block number.

Reply:
test case modified to cover it

Reply:
The test cases include the new 3 digit names but, they do not close and re-open the block catalog and therefore wouldn't trigger the behavior I'm concerned with.  Can we add a test that writes out the segmented blocks log (2 and 3 digit spans), then closes that controller and re-opens a new tester with the same block log directory and ensure that the expected blocks can be read AND have the proper block number after they are read?

Reply:
test case added in test_split_log_replay

Reply:
@b1bart We cannot ensure we test this optimization directly, since we don't know the order the file system will decide to give it to us. The iterator will be invalidated by subsequent file additions if they are earlier, but then active_index will be reset and we will drop the invalid iterator without using it.